File: typegraph/python/typegraph/runtimes/wasm.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0
from dataclasses import dataclass
from typing import List, Optional

from typegraph.gen.exports.runtimes import (
    BaseMaterializer,
    Effect,
    EffectRead,
    MaterializerWasm,
)
from typegraph.gen.types import Err
from typegraph.runtimes.base import Materializer, Runtime
from typegraph.wit import runtimes, store

from typegraph import t


class WasmRuntime(Runtime):
    def __init__(self):
        super().__init__(runtimes.register_wasm_runtime(store))

    def from_wasm(
        self,
        inp: "t.struct",
        out: "t.typedef",
        *,
        func: str,
        wasm: str,
        effect: Optional[Effect] = None,
    ):
        effect = effect or EffectRead()

        mat_id = runtimes.from_wasm_module(
            store,
            BaseMaterializer(runtime=self.id.value, effect=effect),
            MaterializerWasm(wasm_artifact=wasm, func_name=func),
        )

        if isinstance(mat_id, Err):
            raise Exception(mat_id.value)

        return t.func(
            inp,
            out,
            WasmMat(id=mat_id.value, module=wasm, func_name=func, effect=effect),
        )


@dataclass
class WasmMat(Materializer):
    module: str
    func_name: str
    effect: List[str]

File: src/typegraph/deno/src/runtimes/wasm.ts
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

import * as t from "../types.ts";
import { runtimes } from "../wit.ts";
import { Effect } from "../gen/typegraph_core.d.ts";
import { Materializer, Runtime } from "./mod.ts";
import { fx } from "../index.ts";

export class WasmRuntime extends Runtime {
  /** create reflected wasm runtime */
  static reflected(modulePath: string): WasmRuntimeReflected {
    return new WasmRuntimeReflected(modulePath);
  }
  /** create a wasm runtime using the wire protocol */
  static wire(modulePath: string): WasmRuntimeWire {
    return new WasmRuntimeWire(modulePath);
  }
}

interface ReflectedFuncWasmMat extends Materializer {
  funcName: string;
  effect: Effect;
}

interface WireHandlerWasmMat extends Materializer {
  funcName: string;
  effect: Effect;
}

class WasmRuntimeWire extends WasmRuntime {
  constructor(artifactPath: string) {
    super(
      runtimes.registerWasmWireRuntime({
        wasmArtifact: artifactPath,
      }),
    );
  }

  handler<I extends t.Typedef = t.Typedef, O extends t.Typedef = t.Typedef>(
    inp: I,
    out: O,
    {
      name,
      effect = fx.read(),
    }: {
      name: string;
      effect?: Effect;
    },
  ): t.Func<I, O, WireHandlerWasmMat> {
    const matId = runtimes.fromWasmWireHandler(
      {
        runtime: this._id,
        effect,
      },
      {
        funcName: name,
      },
    );

    return t.func(inp, out, {
      _id: matId,
      effect,
      funcName: name,
    });
  }
}

class WasmRuntimeReflected extends WasmRuntime {
  constructor(artifactPath: string) {
    super(
      runtimes.registerWasmReflectedRuntime({
        wasmArtifact: artifactPath,
      }),
    );
  }

  export<I extends t.Typedef = t.Typedef, O extends t.Typedef = t.Typedef>(
    inp: I,
    out: O,
    {
      name,
      effect = fx.read(),
    }: {
      name: string;
      effect?: Effect;
    },
  ): t.Func<I, O, ReflectedFuncWasmMat> {
    const matId = runtimes.fromWasmReflectedFunc(
      {
        runtime: this._id,
        effect,
      },
      {
        funcName: name,
      },
    );

    return t.func(inp, out, {
      _id: matId,
      effect,
      funcName: name,
    });
  }
}
